### PR TITLE
APD: 1560 Check granule exists in ASF Archive

### PR DESCRIPTION
**Pull request by Jlrine2**

### CODE REVIEW

File: api/tests/conftest.py [hunk 1]
```diff
@@ -84,5 +89,16 @@ def make_db_record(job_id,
...
 
 
+def setup_requests_mock(batch):
+    granules = [job['job_parameters']['granule'] for job in batch]
+    cmr_response = {
+        'feed':
+            {
+                'entry': [{'producer_granule_id': granule} for granule in granules]
+            }
+    }
```
Comment (jhkennedy) [inline api/tests/conftest.py:99]: Only thing I see, and is rather nitpick-y, is I'd probably write this like:
```py
     cmr_response = {
        'feed': {
            'entry': [{'producer_granule_id': granule} for granule in granules]
        }
    }
```

File: api/requirements.txt [hunk 2]
```diff
@@ -1,11 +1,12 @@
...
+connexion[swagger-ui]
+cryptography
 flask_api
+Flask-Cors
+moto
 PyJWT
-cryptography
+pytest
+PyYAML
+responses
```
Comment (asjohnston-asf) [inline api/requirements.txt:10]: need to add

File: api/src/hyp3_api/handlers.py [hunk 3]
```diff
@@ -34,6 +39,13 @@ def post_jobs(body, user):
     except QuotaError as e:
         return problem(400, 'Bad Request', str(e))
 
+    try:
+        check_granules_exists([job['job_parameters']['granule'] for job in body['jobs']])
```
Comment (asjohnston-asf) [inline api/src/hyp3_api/handlers.py:43]: could include ` if job['job_type'] == 'RTC_GAMMA'`, not strictly necessary yet though

Comment (asjohnston-asf) [reply api/src/hyp3_api/handlers.py:43]: and maybe break up the creation of the list and the function call:
```
granules = ....
check_granules_exist(granules)
```

File: api/src/hyp3_api/handlers.py [hunk 4]
```diff
@@ -102,6 +114,23 @@ def get_request_time_expression(start, end):
         return key.lte(formatted_end)
 
 
+def check_granules_exists(granules):
```
Comment (asjohnston-asf) [inline api/src/hyp3_api/handlers.py:117]: `check_granules_exist` without the `s` at the end

File: api/src/hyp3_api/handlers.py [hunk 5]
```diff
@@ -102,6 +114,23 @@ def get_request_time_expression(start, end):
...
+        'provider': 'ASF',
+        'collection_concept_ids': {
+            "C1214470488-ASF",  # SENTINEL-1A_SLC
+            "C1327985661-ASF",  # SENTINEL-1B_SLC
+        }
+    }
+    response = requests.get('https://cmr.earthdata.nasa.gov/search/granules.json', params=cmr_parameters)
+    response.raise_for_status()
+    found_granules = [entry['producer_granule_id'] for entry in response.json()['feed']['entry']]
+    if set(granules) != set(found_granules):
```
Comment (asjohnston-asf) [inline api/src/hyp3_api/handlers.py:129]: you can use the subtraction operator on sets to the the difference, https://docs.python.org/2/library/sets.html#set-objects
```
not_found_granules = granules - found_granules
if not_found_granules:
    raise CmrError(f'Requested scenes could not be found: {",".join(not_found_granules)}')
```

File: api/tests/conftest.py [hunk 6]
```diff
@@ -12,6 +15,8 @@
 DEFAULT_JOB_ID = 'myJobId'
 DEFAULT_USERNAME = 'test_username'
 
+CMR_URL_RE = re.compile('https://cmr.earthdata.nasa.gov/search/granules.json.*')
```
Comment (asjohnston-asf) [inline api/tests/conftest.py:18]: might put `CMR_URL = 'https://cmr.earthdata.nasa.gov/search/granules.json'` in `__init__.py` then import it in both `handlers.py` and `conftest.py`

File: api/tests/test_submit_job.py [hunk 7]
```diff
@@ -70,5 +77,25 @@ def test_submit_job_with_empty_description(client):
...
+
+    login(client)
+    response = submit_batch(client, batch)
+    assert response.status_code == status.HTTP_200_OK
+
+    response = submit_batch(client, [make_job('S1A_IW_SLC__1SDV_20200610T173646_20200610T173704_032958_03D14C_5F2A')])
+    assert response.status_code == status.HTTP_400_BAD_REQUEST
+    assert response.json['title'] == 'Bad Request'
+    assert response.json['detail'] == 'Requested scenes could not be found: ' \
+                                      'S1A_IW_SLC__1SDV_20200610T173646_20200610T173704_032958_03D14C_5F2A'
```
Comment (asjohnston-asf) [inline api/tests/test_submit_job.py:101]: Think I'd combine these into a single request with some good granules and some bad granules to verify the error message only complains about the bad ones.  The `200` test doesn't add anything on its own over other test cases for submitting jobs.

File: api/tests/test_submit_job.py [hunk 8]
```diff
@@ -70,5 +77,25 @@ def test_submit_job_with_empty_description(client):
     batch = [
         make_job(description='')
     ]
+    setup_requests_mock(batch)
     response = submit_batch(client, batch)
     assert response.status_code == status.HTTP_400_BAD_REQUEST
+
+
+@responses.activate
```
Comment (asjohnston-asf) [inline api/tests/test_submit_job.py:85]: why is this the only test case with a decorator?

**Comment by asjohnston-asf** [comment]
`handlers.py` is starting to feel a little bloated, might move the helper functions elsewhere and just leave the endpoint functions (similar to what we did with flask views.py)

File: api/tests/test_submit_job.py [hunk 9]
```diff
@@ -70,5 +77,25 @@ def test_submit_job_with_empty_description(client):
...
+    setup_requests_mock(batch)
+
+    login(client)
+    response = submit_batch(client, batch)
+    assert response.status_code == status.HTTP_200_OK
+
+    response = submit_batch(client, [make_job('S1A_IW_SLC__1SDV_20200610T173646_20200610T173704_032958_03D14C_5F2A')])
+    assert response.status_code == status.HTTP_400_BAD_REQUEST
+    assert response.json['title'] == 'Bad Request'
+    assert response.json['detail'] == 'Requested scenes could not be found: ' \
```
Comment (asjohnston-asf) [inline api/tests/test_submit_job.py:100]: `Some requested scenes...` ?

**Pull request merged**